tempesta_fw/t/functional/leaks/test_leak.py

```diff
@@ -0,0 +1,138 @@
+        kmemleakfile = "/sys/kernel/debug/kmemleak"
+        tf_cfg.dbg(2, "Reading kmemleak")
+        if not self.file_exists(kmemleakfile):
+            tf_cfg.dbg(1, "kmemleak file does not exists")
+            return -1
+        cmd = "cat %s | grep \"unreferenced object\" | wc -l" % kmemleakfile
+        [stdout, stderr] = self.tempesta.node.run_cmd(cmd)
+        return int(stdout)
+
+    def get_memory_line(self, name):
+        """ Get value from /proc/meminfo """
+        [stdout, stderr] = self.tempesta.node.run_cmd("cat /proc/meminfo")
+        line = re.search("%s:[ ]+([0-9]+)" % name, stdout)
+        if len(line.groups()) == 0:
+            return -1
+        return int(line.group(1))
+
+    def slab_memory(self):
+        """ Get amount of slab used memory """
+        self.drop_caches()
+        slabmem = self.get_memory_line("Slab")
+        return slabmem
+
+    def free_memory(self):
+        """ Measure total memory usage """
+        self.drop_caches()
+        freemem = self.get_memory_line("MemFree")
+        return freemem
+
+    def assert_tempesta(self):
+        """ We don't check that traffic is parsed correctly. Only detect leaks. """
+
+    def assert_clients(self):
+        """ Check only traffic size. We don't need to check responses. """
+        cl_req_cnt = 0
+        cl_conn_cnt = 0
+        for client in self.clients:
+            req, err = client.results()
+            cl_req_cnt += req
+            cl_conn_cnt += client.connections * self.pipelined_req
+        exp_min = cl_req_cnt
+        exp_max = cl_req_cnt + cl_conn_cnt
+        self.assertTrue(
+            self.tempesta.stats.cl_msg_received >= exp_min and
+            self.tempesta.stats.cl_msg_received <= exp_max
+        )
+
+    def create_clients(self):
+        """ Override to set desired list of benchmarks and their options. """
+        self.clients = [control.Wrk()]
+
+    def create_servers(self):
+        """ Overrirde to create needed amount of upstream servers. """
+        port = tempesta.upstream_port_start_from()
+        self.servers = [control.Nginx(listen_port=port)]
+        for server in self.servers:
+            server.config.set_ka(self.backend_connections)
+
+    def test_kmemleak(self):
+        """ Detecting leaks with kmemleak """
+        kml1 = self.read_kmemleaks()
+        if kml1 == -1:
+            tf_cfg.dbg(1, "No kmemleak detected")
+            return
+        self.generic_test_routine(self.config)
+        kml2 = self.read_kmemleaks()
+        if kml2 == -1:
+            tf_cfg.dbg(1, "No kmemleak detected")
+            return
+        self.assertTrue(kml1 == kml2)
+
+    def test_slab_memory(self):
+        """ Detecting leaks with slab memory measure """
+        used1 = self.slab_memory()
+        if used1 == -1:
+            tf_cfg.dbg(1, "slab memory can't be measured")
+            return
+        self.generic_test_routine(self.config)
+        self.tempesta.stop()
```

Comment:
Ah, I see. You save state of TempestaFW in order not to stop already stopped system during `tearDown()`. It's also possible to set `self.tempesta = None` for the same aim. I believe, we should also stop all servers before measuring free memory since servers can be started on the same node.